v2/federation/watch.go

```diff
@@ -20,16 +26,50 @@ type BundleWatcher interface {
 
 	// OnUpdate is called when a bundle has been updated. If a bundle is
 	// fetched but has not changed from the previously fetched bundle, OnUpdate
-	// will not be called.
+	// will not be called. This function is called synchronously from WatchBundle
+	// so all the time this function takes to finish is added to the time taken
+	// by fetching the bundle.
 	OnUpdate(*spiffebundle.Bundle)
 
 	// OnError is called if there is an error fetching the bundle from the
-	// endpoint.
+	// endpoint. This function is called synchronously from WatchBundle
+	// so all the time this function takes to finish is added to the time taken
+	// by fetching the bundle.
 	OnError(err error)
 }
 
 // WatchBundle watches a bundle on a bundle endpoint. It returns when the
 // context is canceled, returning ctx.Err().
 func WatchBundle(ctx context.Context, trustDomain spiffeid.TrustDomain, url string, watcher BundleWatcher, options ...FetchOption) error {
-	panic("not implemented")
+	if watcher == nil {
+		return federationErr.New("watcher cannot be nil")
+	}
+
+	var lastBundle *spiffebundle.Bundle
+	d := watcher.NextRefresh(0)
+	timer := time.NewTimer(d)
```

Comment:
I don't think the timer should be running while the call to fetch bundle is happening. I'd suggest something like:

```
var timer *time.Timer
for {
    bundle, err = fetchBundle...
    // handle error, do equality check and set lastBundle

    var nextRefresh time.Duration
    if refreshHint, ok := latestBundle.RefreshHint(); ok {
        nextRefresh = watcher.NextRefresh(refreshHint)
    }

    if timer == nil {
        timer = time.NewTimer(nextRefresh)
        defer timer.Stop()
    } else {
        timer.Reset(refreshHint)    
    }

    select {
        case <-timer.C:
        case <-ctx.Done():
            return nil, ctx.Err()
    }
}
```
}

Reply:
What if latestBundle.RefreshHint() returns !ok, nextRefresh would be 0, resulting in no delay before the next bundle fetching. So I was wondering if we should do this instead:
```go
    var nextRefresh time.Duration
    if refreshHint, ok := latestBundle.RefreshHint(); ok {
        nextRefresh = watcher.NextRefresh(refreshHint)
    } else {
        nextRefresh = watcher.NextRefresh(0)
    }
```

Reply:
The watcher provides the next refresh period. It is passed the hint, if available, but ultimately it is up to the watcher to decide how often it wants to refresh. From the Watcher interface docs

```
	// NextRefresh is called by Watch to determine when the next refresh
	// should take place. A refresh hint is provided, which can be zero, meaning
	// the watcher is free to choose its own refresh cadence. If the refresh hint
	// is non-zero, the watcher SHOULD return a next refresh time at or below
	// that to ensure the bundle stays up-to-date.
	NextRefresh(refreshHint time.Duration) time.Duration
```

Reply:
Hmm, we should update that to say `If the refresh hint is greater than zero`, instead of `non-zero`.

Reply:
Oh, I see where my sample was wrong. Yes, that looks good.